// langchain@0.0.151/dist/retrievers/multi_query.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MultiQueryRetriever = void 0;
const llm_chain_js_1 = require("../chains/llm_chain.cjs");
const prompt_js_1 = require("../prompts/prompt.cjs");
const output_parser_js_1 = require("../schema/output_parser.cjs");
const retriever_js_1 = require("../schema/retriever.cjs");
class LineListOutputParser extends output_parser_js_1.BaseOutputParser {
    constructor() {
        super(...arguments);
        Object.defineProperty(this, "lc_namespace", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: ["langchain", "retrievers", "multiquery"]
        });
    }
    static lc_name() {
        return "LineListOutputParser";
    }
    async parse(text) {
        const startKeyIndex = text.indexOf("<questions>");
        const endKeyIndex = text.indexOf("</questions>");
        const questionsStartIndex = startKeyIndex === -1 ? 0 : startKeyIndex + "<questions>".length;
        const questionsEndIndex = endKeyIndex === -1 ? text.length : endKeyIndex;
        const lines = text
            .slice(questionsStartIndex, questionsEndIndex)
            .trim()
            .split("\n")
            .filter((line) => line.trim() !== "");
        return { lines };
    }
    getFormatInstructions() {
        throw new Error("Not implemented.");
    }
}
// Create template
const DEFAULT_QUERY_PROMPT = /* #__PURE__ */ new prompt_js_1.PromptTemplate({
    inputVariables: ["question", "queryCount"],
    template: `You are an AI language model assistant. Your task is
to generate {queryCount} different versions of the given user
question to retrieve relevant documents from a vector database.
By generating multiple perspectives on the user question,
your goal is to help the user overcome some of the limitations
of distance-based similarity search.

Provide these alternative questions separated by newlines between XML tags. For example:

<questions>
Question 1
Question 2
Question 3
</questions>

Original question: {question}`,
});
// Export class
class MultiQueryRetriever extends retriever_js_1.BaseRetriever {
    static lc_name() {
        return "MultiQueryRetriever";
    }
    constructor(fields) {
        super(fields);
        Object.defineProperty(this, "lc_namespace", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: ["langchain", "retrievers", "multiquery"]
        });
        Object.defineProperty(this, "retriever", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "llmChain", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "queryCount", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 3
        });
        Object.defineProperty(this, "parserKey", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "lines"
        });
        this.retriever = fields.retriever;
        this.llmChain = fields.llmChain;
        this.queryCount = fields.queryCount ?? this.queryCount;
        this.parserKey = fields.parserKey ?? this.parserKey;
    }
    static fromLLM(fields) {
        const { retriever, llm, prompt = DEFAULT_QUERY_PROMPT, queryCount, parserKey, ...rest } = fields;
        const outputParser = new LineListOutputParser();
        const llmChain = new llm_chain_js_1.LLMChain({ llm, prompt, outputParser });
        return new this({ retriever, llmChain, queryCount, parserKey, ...rest });
    }
    // Generate the different queries for each retrieval, using our llmChain
    async _generateQueries(question, runManager) {
        const response = await this.llmChain.call({ question, queryCount: this.queryCount }, runManager?.getChild());
        const lines = response.text[this.parserKey] || [];
        if (this.verbose) {
            console.log(`Generated queries: ${lines}`);
        }
        return lines;
    }
    // Retrieve documents using the original retriever
    async _retrieveDocuments(queries, runManager) {
        const documents = [];
        for (const query of queries) {
            const docs = await this.retriever.getRelevantDocuments(query, runManager?.getChild());
            documents.push(...docs);
        }
        return documents;
    }
    // Deduplicate the documents that were returned in multiple retrievals
    _uniqueUnion(documents) {
        const uniqueDocumentsDict = {};
        for (const doc of documents) {
            const key = `${doc.pageContent}:${JSON.stringify(Object.entries(doc.metadata).sort())}`;
            uniqueDocumentsDict[key] = doc;
        }
        const uniqueDocuments = Object.values(uniqueDocumentsDict);
        return uniqueDocuments;
    }
    async _getRelevantDocuments(question, runManager) {
        const queries = await this._generateQueries(question, runManager);
        const documents = await this._retrieveDocuments(queries, runManager);
        const uniqueDocuments = this._uniqueUnion(documents);
        return uniqueDocuments;
    }
}
exports.MultiQueryRetriever = MultiQueryRetriever;
